### ISSUE TITLE
Cannot retry a transaction after a deadlock

### ISSUE THREAD

**Issue by tzfrs** [doctrine/DoctrineBundle#865]
Hello,

we have an API endpoint that adds an item to the cart. A user can add all items from a previous order to the cart, meaning the endpoint is called 50 times. Since all 50 calls update a cart, this can lead to a deadlock. What I'm trying now is to retry the transaction, however it doesn't work.

Doesn't work means that the entity is not flushed into the database. I'm adding 50 products, one product get's the deadlock error and runs into the `catch` block of the code below and is never saved into the database.

This is the code

```
$manager->persist($cartItem);
$successfullyFlushed = false;
$retry = 1;
$maxRetry = 5;

while (!$successfullyFlushed && $retry <= $maxRetry) {
    try {
        $manager->flush();
        $successfullyFlushed = true;
    } catch (DriverException $exception) {
        if (!$manager->isOpen()) {
            $manager = $manager->create($manager->getConnection(), $manager->getConfiguration());
        }

        if ($exception->getErrorCode() === 1213) {
            if ($retry === $maxRetry) {
                throw  $exception;
            }

            continue;
        }

        throw $exception;
    }
}
```

I tried many things, such as calling `beginTransaction`, `commit` and `rollBack` manually like explained in here: https://stackoverflow.com/a/13150683/2823419

But none of these yielded results. I also called `detach($cartItem)` in the catch block and persisted at again, but I got an error for that as well.

So my question is: What is the "correct" way to retry a `flush` after a deadlock?

**Comment by stof** [comment]
Well, if the manager gets closed, you create a new one in yoru code. But this new manager is *not* managing any of your objects.

**Comment by tzfrs** [comment]
Hi @stof,

yes, that's sth. we thought of as well, so what I did was do persist the `entity` again by doing

`$manager->persist($cartItem)` right before the flush instead of before the `while` loop so it looked like this

```
// ... Previous code

while (!$successfullyFlushed && $retry <= $maxRetry) {
    try {
        $manager->persist($cartItem)
        $manager->flush();

       // .... Rest of code
```

 however this threw an error:

> Doctrine\ORM\ORMInvalidArgumentException: "A new entity was found through the relationship '{Namespace}\AbstractItem#product' that was not configured to cascade persist operations for entity: {Namespace}\\Product@000000001b9ce8a100000000271b2c0d.

which it only does when a deadlock happens. What I'm doing before and isn't shown in the code sample above is the following code

`$product = $productRepository->findOneBy($itemNumber, $vendorToken);`

Could this be also related to the fact that the `EntityManager` was closed because of the deadlock and means that it doesn't have the `Product` anymore as well? Does this mean, that whenever I have a deadlock and the `EntityManager` closes, I have to do everything database related again?

**Comment by Ocramius** [comment]
Correct, anything you loaded (which is in "managed" state) needs to either
be registered with the new EntityManager, or fetched from it again.

**Comment by tzfrs** [comment]
Okay, thank you for your answer, but isn't that kind of an overkill? I just did this and stumbled upon the next problem, because we're using proxies, so when retrieving the cart again and checking if the user is `authorized` using `$cart->getUser() === $this->getUser()` it returns false, so I always have to use a real object in there or change the check to `getId()`. And even then I get the same error that `product` is not persisted. And I'm sure that even if I fix those problems, other's will occur and it seems like I have to adjust the whole application just to retry a simple transaction.

I mean I can understand why the `EntityManager` closes on error, however for a Deadlock MySQL (for example) suggests to just retry the transaction, so wouldn't it make sense to differentiate between a deadlock and any other error and in case of deadlock not close the `EntityManager` at all?

I saw this PR https://github.com/doctrine/doctrine2/pull/806 which was closed (due to bad code?) but isn't this generally a good idea? To retry this? Maybe make the number of retrys configurable and set `default` to `1` so everyone updating will not even have a bc break?

**Comment by Ocramius** [comment]
Once an interaction failed, the state of the UnitOfWork cannot be relied
upon anymore, hence why the EntityManager is being closed.

This is a conscious design decision that won't be changed, not even for 3.x.

An example is with post-insert Identifiers: what to do after you've
assigned a bunch of them, and a statement fails? Think about just that and
its performance and coding implications.

**Comment by tzfrs** [comment]
Okay @Ocramius @stof .

A co-worker found a very old GitHub issue https://github.com/symfony/symfony/issues/5339#issuecomment-8022779

When I exchanged 

```php
if (!$manager->isOpen()) {
    $manager = $manager->create($manager->getConnection(), $manager->getConfiguration());
}
```

with

```php
if (!$manager->isOpen()) {
    $manager = $this->getDoctrine()->resetManager();
}
```

everything worked as expected. So it seemed like I had to `reset` the manager instead of just creating a new one.

Do you see any problems with this solution?

**Comment by Majkl578** [comment]
Yes, the new resetted manager will hold (manage) no entities.

This is a duplicate of https://github.com/doctrine/doctrine2/issues/6272.

**Comment by stof** [comment]
that won't change anything, as `resetManager()` will recreate a new EntityManager too (the difference is that it will also replace the manager in the container, not only in your local variable)

**Comment by Majkl578** [comment]
`resetManager()` is out of scope both here in DoctrineBundle repo and ORM repo - it's part of Symfony's Doctrine bridge.
In ORM 3.0 EM may become resettable directly (https://github.com/doctrine/doctrine2/issues/5933).

**Comment by tzfrs** [comment]
Now I'm confused. When using `resetManager` everything works as expected and the cart item is persisted without any errors, however your (@stof / @Majkl578 ) comments make it sound like this shouldn't be the case?

**Comment by stof** [comment]
@Majkl578 if your repositories are defined as a service, resetting the manager is necessary for things to work, otherwise the repository will still use the old manager, not the only you recreated manually without telling DoctrineBundle about it.
